### DISCUSSION TITLE
Little help with m5stack

### DISCUSSION THREAD

**Discussion by hcl-83** [vortigont/pzem-edl#6] [General]
Hello,
i want to use pzem-003 DC.
I use m5stack tough with UIFlow (block programming / micropython)

So i need simple just send request and get respods from pzem-003.

Can You help me what to send and how to read it.
My project is just read values so no speed is needed i can send request wait 2s and read but i dont know what to send and how read.
I try simple 
uart1 = machine.UART(1, tx=14, rx=13)
uart1.init(9600, bits=8, parity=None, stop=2)
uart1.write (8bit like 0x02 ...) 
and after 2sek uart1.any() but not get any bytes to read :/

Regards Kamil :>

**Comment by vortigont** [comment]
Hi! Sorry, but this library won't work with micropython.  It runs under ISP-IDF/Arduino. You can check [examples](https://github.com/vortigont/pzem-edl/tree/main/examples) section for some Arduino's sketches to start with.
If you want to implement your own methods on python you can find [datasheet](https://github.com/vortigont/pzem-edl/blob/main/docs/PZEM-003-Manual.pdf) for PZEM-003 for available command codes to send/read via serial.

**Comment by hcl-83** [comment]
could you direct me because I am not proficient in uart / rs485 how the request to the sensor should look like these 8 bits because I know the syntax but I do not know what to send to receive data Volts Amperes and Power.
I dont know how to read this datasheet so dont know what to send :)

**Comment by vortigont** [comment]
you do not expect me to write a new lib for you, right? :))) Better start with learning how to read datasheets before attempting to program a device it refers to.
BTW, you could use some existing micropython lib like [pidou46/uPZEM-004t](https://github.com/pidou46/uPZEM-004t)

**Comment by vortigont** [comment]
it is pretty straight forward written in datasheet

```
For example, the master sends the following command (CRC check code is replaced by
0xHH and 0xLL, the same below):
0x01 + 0x04 + 0x00 + 0x00 + 0x00 + 0x08 + 0xHH + 0xLL
Indicates that the master needs to read 8 registers with slave address 0x01 and the start
address of the register is 0x0000.
The correct reply from the slave is as following:
0x01 + 0x04 + 0x10 + 0x27 + 0x10 + 0x00 + 0x64 + 0x03 + 0xE8 + 0x00 + 0x00 + 0x00 +
0x00 + 0x00 + 0x00 + 0x00 + 0x00 + 0x00 + 0x00 + 0xHH + 0xLL
The above data shows
⚫ Voltage is 0x2710，converted to decimal is 10000,display 100.00V；
⚫ Current is 0x0064，converted to decimal is 100,display 1.00A；
⚫ Power is 0x000003E8，converted to decimal is 1000,display 100.0W；
⚫ Energy is 0x00000000，converted to decimal is 0,display 0Wh；
⚫ High voltage alarm status 0x0000,indicates the current voltage is lower than the high
voltage threshold.
⚫ Low voltage alarm status 0x0000,indicates the current voltage is higher than the low
voltage threshold.
```

**Comment by hcl-83** [comment]
of course I don't require you to write a library :) for me. Thank you very much for responding. I can send a query but I don't know what to enter, what 8 bit hex. Manual says: x01 + 0x04 + 0x00 + 0x00 + 0x00 + 0x08 + 0xHH + 0xLL but I can't type 0xHH + 0xLL. These values ​​mean something, but I don't know what. Could you give me one ready-made 8-bit query, I'll try to deduce on this basis how it is generated :) and see if the sensor will answer me at all ;p Regards, Kamil

**Comment by vortigont** [comment]
`0xHH + 0xLL ` is a two byte modbus message checksum. It must be algorithmically calculated for each unique request. I can't provide you one out of my mind.
I recommend you to dedicate a bit of your efforts to read the datasheet and understand how it works. You won't be able to implement it via deducing bytes.
Try to build and run an example of my library with enabled debugging, there you can find req/reply bytes dumps if you need.

**Comment by hcl-83** [comment]
I read about crc last night so now i test micropython script for calculated checksum.

Big Thanx for YOU :> and Happy New Year
